File: CoreSkills/Graphs/dijkstra_shortestpath.cpp
#include <unordered_map>
#include <vector>
#include <tuple>
#include <queue>
#include <unordered_set>

class Solution {
public:
    std::unordered_map<int, int> shortestPath(int n, std::vector<std::vector<int>>& edges, int src) {
        /*
        - return the shortest distance from the starting vertex to every other vertex
        - dijkstra's algorithm: greedy algo - put edges on a minheap and pop off edges with min total cost
        - if a vertex is unreachable it's shortest path distance is -1

        Input:
        n = 5 (number of vertices)
        edges = [[0,1,10], [0,2,3], [1,3,2], [2,1,4], [2,3,8], [2,4,2], [3,4,5]] (u, v, w) (start, dst, weight)
        src = 0 (starting vertex)

        Output:
        {{0:0}, {1:7}, {2:3}, {3:9}, {4:5}}

        */

        std::unordered_map<int,int> shortestpaths; //key: vertex, value: shortest path distance
        for (int i = 0; i < n; i++) {
            shortestpaths[i] = -1;
        }

        // 1. Building adjacency list
        std::unordered_map<int, std::vector<std::tuple<int,int>>> graph;
        for (int i = 0; i < edges.size(); i++) {
            int v = edges[i][0]; // src
            int u = edges[i][1]; // dst
            int w = edges[i][2]; // weight
            graph[v].push_back(std::make_tuple(w, u));
        }

        // 2. Setup dijkstra's
        std::unordered_set<int> visited;
        std::priority_queue<std::tuple<int,int>> minheap; // tuple(total cost, dst vertex), first element of tuple is treated as priority
        // priority queues are maxheaps by default -> negate weight of edges to achieve a minheap
        visited.insert(src);
        for (int i = 0; i < graph[src].size(); i++) {
            // push the outgoing edges from src onto the minheap
            int weight = get<0>(graph[src][i]);
            int dst = get<1>(graph[src][i]);
            minheap.emplace(-weight, dst); // or could minheap.push(std::make_tuple())
        }
        shortestpaths[src] = 0;

        // 3. execute dijkstra's
        while(!minheap.empty()) {
            std::tuple<int,int> edge = minheap.top();
            int weight = get<0>(edge);
            int dst = get<1>(edge);
            minheap.pop();
            weight = weight * -1; // convert back to positive weight

            // if the dst is not visited, update shortestpaths
            if (visited.find(dst) == visited.end()) {
                visited.insert(dst);
                shortestpaths[dst] = weight;
            }

            // add edges from dst to unvisited vertices
            for (std::tuple<int,int> edge : graph[dst]) { // looping through the outgoing edges of dst vertex
                int w = get<0>(edge);
                int v = get<1>(edge);
                if (visited.find(v) == visited.end()) {
                    minheap.emplace(-1*(w + shortestpaths[dst]), v);
                }
            }
        }

        return shortestpaths;

        // Time Complexity: O(ElogE) because for E edges we have to push and pop them from the minheap which takes logE time.





        


        
    }
};

File: CoreSkills/Graphs/dijkstra_shortestpath.py
import heapq

class Solution:
    def shortestPath(self, n: int, edges: list[list[int]], src: int) -> dict[int, int]:
        """
        - implement dijkstra's shortest path
        - input: weighted, directed graph and starting vertex
        - output: shortest distance from the start to every other vertex
        - if a vertex is unreachable from the source vertex, the shortest path distance is -1

        Input:
        n = 5 (number of vertices)
        edges = [[0,1,10], [0,2,3], [1,3,2], [2,1,4], [2,3,8], [2,4,2], [3,4,5]] (u, v, w) (edge from u to v with weight w)
        src = 0 (starting vertex)

        Output:
        {{0:0}, {1:7}, {2:3}, {3:9}, {4:5}} (key: vertex, value: shortest path)
        """
        # 1. Build adjacency list & output
        graph = {}
        shortestpaths = {}
        for i in range(n):
            graph[i] = [] # edges will be represented as [dst vertex, weight]
            shortestpaths[i] = -1
        # add edges to adjacency list
        for e in edges:
            u = e[0]
            v = e[1]
            w = e[2]
            graph[u].append([v,w])
        
        # 2. Setup Dijkstra's
        visited = set()
        visited.add(src) # mark src as visited
        minheap = []
        shortestpaths[src] = 0
        for edge in graph[src]:
            # add outgoing edges from the source to minheap
            heapq.heappush(minheap, (edge[1], edge[0])) # priority is determeined with respect to first element (cost)
            #(cost, dst)

        # 3. Execute Dijkstra's
        while len(minheap) > 0:
            edge = heapq.heappop(minheap) # min cost edge
            totalcost = edge[0]
            dst = edge[1]

            # update shortest path
            if dst not in visited:
                visited.add(dst)
                shortestpaths[dst] = totalcost
            
            # add adjacent edges
            for e in graph[dst]:
                v = e[0]
                w = e[1]
                if v not in visited: # if dst vertex unvisited (don't know shortest path)
                    heapq.heappush(minheap, (w + shortestpaths[dst], v))
        
        return shortestpaths

"""
Time Complexity? Basically we have E edges and for each edge we will push and pop from the minheap.
Pushing and popping takes logE because we could have E edges on the minheap so our time complexity is O(E*logE)
"""


            






        

    

